server/v2/mempool/mempool.go

```diff
@@ -2,22 +2,35 @@ package mempool
 
 import (
 	"context"
-)
 
-// Mempool defines the required methods of an application's mempool.
-type Mempool[T any] interface {
-	// Insert attempts to insert a Tx into the app-side mempool returning
-	// an error upon failure. Insert will validate the transaction using the txValidator
-	Insert(ctx context.Context, txs T) error
+	"cosmossdk.io/server/v2/core/transaction"
+)
 
-	// GetTxs returns a list of transactions to add in a block
-	// size specifies the size of the block left for transactions
-	GetTxs(ctx context.Context, size uint32) (ts any, err error)
+func NewCacheTx[T transaction.Tx](tx T, bytes []byte) CacheTx[T] {
+	return CacheTx[T]{
+		Decoded: tx,
+		Encoded: bytes,
+	}
+}
 
-	// CountTx returns the number of transactions currently in the mempool.
-	CountTx() uint32
+// CacheTx defines a transaction.Tx containing both encoded and decoded version of it.
+// This is useful to avoid further encoding and decoding of txs in the process
+// of block building.
+type CacheTx[T transaction.Tx] struct {
+	// Decoded is the decoded Tx.
+	Decoded T
+	// Encoded is the encoded Tx.
+	Encoded []byte
+}
 
-	// Remove attempts to remove a transaction from the mempool, returning an error
-	// upon failure.
-	Remove(txs T) error
+// Mempool defines the required methods of an application's mempool.
+type Mempool[T transaction.Tx] interface {
```

Comment:
how do you remove a tx from the mempool?

Reply:
In the mempool, I've added this method called `Pull`, the idea is that during `PrepareBlock` we can `Pull` TXs from the mempool, which removes them. 

Maybe the name `Pull`  is not really good, but the idea is that the mempool is something consumers (clients, consensus) push to, and the AppManager then pops from it and builds the block.

Reply:
discussed in a call, we will keep pull and push. When we pull the transactions consensus has the option of keeping a cache of them, if the proposer is superseeded by another it would push the transactions it pulled back into the mempool

Reply:
Note, you don't want to actually remove txs during `PrepareProposal`, since the proposal can be scraped and a new consensus round proceeds. Ideally, the txs are removed during FinalizeBlock or whenever the block is processed.